exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/DrillFileSystem.java

```diff
@@ -83,28 +87,60 @@
   private final OperatorStats operatorStats;
   private final CompressionCodecFactory codecFactory;
 
+  private boolean initialized = false;
+
   public DrillFileSystem(Configuration fsConf) throws IOException {
     this(fsConf, null);
   }
 
   public DrillFileSystem(Configuration fsConf, OperatorStats operatorStats) throws IOException {
+    Preconditions.checkNotNull(fsConf);
+
+    // Configuration objects are mutable, and the underlying FileSystem object may directly use a passed in Configuration.
+    // In order to avoid scenarios where a Configuration can change after a DrillFileSystem is created, we make a copy
+    // of the Configuration.
+    fsConf = new Configuration(fsConf);
     this.underlyingFs = FileSystem.get(fsConf);
+
+    logger.trace("Configuration for the DrillFileSystem {} {}, underlyingFs: {}",
+      fsConf.getTrimmed(FS_DEFAULT_NAME),
+      fsConf.getTrimmed(FS_DEFAULT_NAME_KEY),
+      this.underlyingFs.getUri());
+
+    if (logger.isTraceEnabled()) {
+      logger.trace("Who made me? {}", new RuntimeException("Who made me?"));
+    }
+
     this.codecFactory = new CompressionCodecFactory(fsConf);
     this.operatorStats = operatorStats;
+    this.initialized = true;
   }
 
+  /**
+   * This method should not be used on the {@link DrillFileSystem}. Once a {@link DrillFileSystem} is created it is immutable.
+   * @throws UnsupportedOperationException
+   * @param conf
+   */
   @Override
   public void setConf(Configuration conf) {
-    // Guard against setConf(null) call that is called as part of superclass constructor (Configured) of the
-    // DrillFileSystem, at which point underlyingFs is null.
-    if (conf != null && underlyingFs != null) {
-      underlyingFs.setConf(conf);
+    if (!initialized) {
+      // The parent class's constructor FileSystem() calls Configured(null) which calls setConf(null).
+      // We want to let that first call to setConf succeed. Any subsequent calls would be made by a user and are not allowed.
+      return;
     }
+
+    // A programmer using this class should never call this method.
+    throw new UnsupportedOperationException(DrillFileSystem.class.getCanonicalName() + " is immutable and should not be changed after creation.");
```

Comment:
Do you have this exception in the process of reproducing the issue with the changes from this PR?

Reply:
This was intentional. The idea is that if a developer makes a change where they call the setter their build will fail, since they violated the contract.

Reply:
It's good. I just was interested how immutability of DFS solves the original issue, where and how DFS was changing?
Is `setConf()` method the main fix of the issue? Maybe this method is called and fall into return statement?